src/util/Tokenizer.ts

```diff
@@ -0,0 +1,43 @@
+export class Tokenizer {
+    private static tokenizerInstance: Tokenizer;
+    private tokens: string[];
+
+    private constructor(fileName: string) {
+        // TODO: implement this
+    }
+
+    public static makeInstance(fileName: string) {
+        if(!this.tokenizerInstance) {
+            this.tokenizerInstance = new Tokenizer(fileName);
+        }
+    }
+
+    public static getInstance(): Tokenizer {
+        return this.tokenizerInstance;
```

Comment:
I'm not sure how we instantiate this class, but how we have it right now means that there's a 2-step process to use the `Tokenizer`
```TypeScript
Tokenizer.makeInstance("foobar");
let tokenizer: Tokenizer = Tokenizer.getInstance();
```

I think we can simplify this by having

```TypeScript
public static getInstance(fileName: string): Tokenizer {
    if (this.tokenizerInstance === null || this.tokenizerInstance === undefined) {
        this.tokenizerInstance = new Tokenizer(fileName);
    }
    return this.tokenizerInstance;
}
```
What are your thoughts? Is there a reason you wanted to have the setup/getInstance in two different methods? I think that would make sense if we were to create multiple instances, as it would decouple setup + instantiation, but I'm not sure if it would make sense here.

Reply:
Thanks for the feedback!

One concern I have is that would require the caller to specify an input string every time we want to getInstance, even though it’s not going to actually be used in most calls to getInstance, and it wouldn’t be clear if we did something like: 

`this.tokenizer.getInstance(“foo”); //creates and returns instance of tokenizer for foo file
this.tokenizer.getIntance(“bar”); //returns instance of tokenizer for bar file`

The current implementation also has a similar fault, in that if I did something like:

`this.tokenizer.makeInstance(“foo”); //creates instance of tokenizer (file = foo)
this.tokenizer.getInstance(); //get instance, file = foo
this.tokenizer.makeInstance(“bar”); //creates instance (file = foo)
this.tokenizer.getInstance(); //get instance, file still = foo`

we’d run into a similar issue. To better handle that issue, I see two options (assuming we stick to this general pattern of getInstance):
1: Add a throw new AssertionError to makeInstance if this.tokenizerInstance is set, leave getInstance as is.
2: Update getInstance as you suggested, but make fileName an optional parameter and add some errorHandling in getInstance, and remove makeInstance

Reply:
I'm not sure if I understand the fault in the current implementation. 

I think having `getInstance()` and `makeInstance(fileName: string)` is fine that you don't have a string parameter every time you call `getInstance()`. But `getInstance()` should throw an exception if `makeInstance()` hasn't been called yet. The 2-step way to use it is only once at the start, otherwise its your usual `getInstance()` call in the ASTNodes.

Reply:
Ah, sorry - that makes a lot more sense. Forgive me for asking, but why do we need the file name at all in our tokenizer? Are we going to use the name to make a call to readFile within the tokenizer class?

Reply:
@jyoo980 exactly- I was thinking that we would want to be able to specify the file that we want to tokenize.
@scveloso updated this to include throwing an error in getInstance() if it's called before Tokenizer is initialized.